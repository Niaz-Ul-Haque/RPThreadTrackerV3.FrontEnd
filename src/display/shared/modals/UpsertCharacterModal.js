import React from 'react';
import PropTypes from 'prop-types';
import { Modal, ModalHeader, ModalBody, ModalFooter, Button } from 'reactstrap';
import { AvForm } from 'availity-reactstrap-validation';
import UpsertCharacterForm from '../../forms/upsert-character/UpsertCharacterForm';
import TooltipForm from '../../forms/TooltipForm';

const propTypes = {
	isUpsertCharacterModalOpen: PropTypes.bool.isRequired,
	submitUpsertCharacter: PropTypes.func.isRequired,
	closeUpsertCharacterModal: PropTypes.func.isRequired,
	characterToEdit: PropTypes.shape({}).isRequired
};

class UpsertCharacterModal extends React.Component {
	constructor(props) {
		super(props);
		this.handleInputChange = this.handleInputChange.bind(this);
		this.state = {
			characterToEdit: props.characterToEdit
		};
	}

	componentWillReceiveProps(nextProps) {
		this.setState({ characterToEdit: nextProps.characterToEdit });
	}

	handleInputChange(event) {
		const { target } = event;
		const value = target.type === 'checkbox' ? target.checked : target.value;
		const { name } = target;
		this.setState({
			characterToEdit: Object.assign({}, this.state.characterToEdit, {
				[name]: value
			})
		});
	}
	render() {
		const {
			isUpsertCharacterModalOpen,
			submitUpsertCharacter,
			closeUpsertCharacterModal,
			characterToEdit
		} = this.props;
		if (!characterToEdit) {
			return (
				<div />
			);
		}
		return (
			<Modal isOpen={isUpsertCharacterModalOpen} toggle={closeUpsertCharacterModal} backdrop>
				<AvForm onValidSubmit={() => submitUpsertCharacter(this.state.characterToEdit)}>
					<ModalHeader toggle={closeUpsertCharacterModal}>{characterToEdit.characterId ? 'Edit Character' : 'Add Character'}</ModalHeader>
					<ModalBody>
						<TooltipForm
							Renderable={UpsertCharacterForm}
							characterToEdit={characterToEdit}
							handleInputChange={this.handleInputChange}
						/>
					</ModalBody>
					<ModalFooter>
						<Button color="primary">Submit Character</Button>{' '}
						<Button color="secondary" onClick={closeUpsertCharacterModal}>Cancel</Button>
					</ModalFooter>
				</AvForm>
			</Modal>
		);
	}
}

UpsertCharacterModal.propTypes = propTypes;
export default UpsertCharacterModal;
