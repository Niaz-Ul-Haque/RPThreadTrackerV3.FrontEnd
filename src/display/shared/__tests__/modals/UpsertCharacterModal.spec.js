// #region imports
import React from 'react';
import { shallow } from 'enzyme';
import { getSpecWrapper } from '../../../../../config/tests/helpers.unit';
import UpsertCharacterModal from '../../modals/UpsertCharacterModal';
// #endregion imports

// #region mocks
jest.mock('../../../forms/upsert-character/UpsertCharacterForm', () => 'UpsertCharacterForm');
jest.mock('../../../forms/TooltipForm', () => 'TooltipForm');
// #endregion mocks

const createTestProps = propOverrides => ({
	isUpsertCharacterModalOpen: true,
	submitUpsertCharacter: jest.fn(),
	closeUpsertCharacterModal: jest.fn(),
	characterToEdit: { characterId: 1 },
	...propOverrides
});

describe('rendering', () => {
	it('should render valid snapshot', () => {
		const props = createTestProps();
		const jsx = (<UpsertCharacterModal {...props} />);
		const element = shallow(jsx);
		expect(element).toMatchSnapshot();
	});
	it('should render valid snapshot if character is new', () => {
		const props = createTestProps({ characterToEdit: {} });
		const jsx = (<UpsertCharacterModal {...props} />);
		const element = shallow(jsx);
		expect(element).toMatchSnapshot();
	});
});

describe('behavior', () => {
	it('should trigger upsert character action on form submit', () => {
		const submitUpsertCharacter = jest.fn();
		const characterToEdit = { characterId: 2 };
		const props = createTestProps({ submitUpsertCharacter, characterToEdit });
		const jsx = (<UpsertCharacterModal {...props} />);
		const element = shallow(jsx);
		const form = getSpecWrapper(element, 'upsert-character-modal-form');
		form.prop('onValidSubmit')();
		expect(submitUpsertCharacter).toHaveBeenCalledTimes(1);
		expect(submitUpsertCharacter).toHaveBeenLastCalledWith(characterToEdit);
	});
	it('should handle input change with text value', () => {
		const submitUpsertCharacter = jest.fn();
		const event = { target: { type: 'text', name: 'characterName', value: 'My Character' } };
		const props = createTestProps({ submitUpsertCharacter });
		const jsx = (<UpsertCharacterModal {...props} />);
		const element = shallow(jsx);
		element.instance().handleInputChange(event);
		const form = getSpecWrapper(element, 'upsert-character-modal-form');
		form.prop('onValidSubmit')();
		expect(submitUpsertCharacter).toHaveBeenCalledTimes(1);
		expect(submitUpsertCharacter).toHaveBeenLastCalledWith({ characterId: 1, characterName: 'My Character' });
	});
	it('should handle input change with checkbox value', () => {
		const submitUpsertCharacter = jest.fn();
		const event = { target: { type: 'checkbox', name: 'isActive', checked: true } };
		const props = createTestProps({ submitUpsertCharacter });
		const jsx = (<UpsertCharacterModal {...props} />);
		const element = shallow(jsx);
		element.instance().handleInputChange(event);
		const form = getSpecWrapper(element, 'upsert-character-modal-form');
		form.prop('onValidSubmit')();
		expect(submitUpsertCharacter).toHaveBeenCalledTimes(1);
		expect(submitUpsertCharacter).toHaveBeenLastCalledWith({ characterId: 1, isActive: true });
	});
	it('should set character to edit on receive props', () => {
		const submitUpsertCharacter = jest.fn();
		const characterToEdit = { characterId: 2 };
		const props = createTestProps({ submitUpsertCharacter });
		const jsx = (<UpsertCharacterModal {...props} />);
		const element = shallow(jsx);
		element.setProps({ characterToEdit });
		const form = getSpecWrapper(element, 'upsert-character-modal-form');
		form.prop('onValidSubmit')();
		expect(submitUpsertCharacter).toHaveBeenCalledTimes(1);
		expect(submitUpsertCharacter).toHaveBeenLastCalledWith(characterToEdit);
	});
	it('should close modal on modal toggle', () => {
		const closeUpsertCharacterModal = jest.fn();
		const props = createTestProps({ closeUpsertCharacterModal });
		const jsx = (<UpsertCharacterModal {...props} />);
		const element = shallow(jsx);
		const header = getSpecWrapper(element, 'upsert-character-modal');
		header.prop('toggle')();
		expect(closeUpsertCharacterModal).toHaveBeenCalledTimes(1);
	});
	it('should close modal on close button click', () => {
		const closeUpsertCharacterModal = jest.fn();
		const props = createTestProps({ closeUpsertCharacterModal });
		const jsx = (<UpsertCharacterModal {...props} />);
		const element = shallow(jsx);
		const button = getSpecWrapper(element, 'upsert-character-modal-close-button');
		button.simulate('click');
		expect(closeUpsertCharacterModal).toHaveBeenCalledTimes(1);
	});
});
