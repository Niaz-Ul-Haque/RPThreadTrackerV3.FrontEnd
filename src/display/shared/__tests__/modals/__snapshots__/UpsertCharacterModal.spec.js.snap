// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering should render valid snapshot 1`] = `
<Modal
  autoFocus={true}
  backdrop={true}
  backdropTransitionTimeout={150}
  data-spec="upsert-character-modal"
  fade={true}
  isOpen={true}
  keyboard={true}
  modalTransitionTimeout={300}
  toggle={[MockFunction]}
  zIndex={1050}
>
  <AvForm
    data-spec="upsert-character-modal-form"
    method="get"
    model={Object {}}
    onInvalidSubmit={[Function]}
    onKeyDown={[Function]}
    onSubmit={[Function]}
    onValidSubmit={[Function]}
    tag={[Function]}
    validationEvent={
      Array [
        "onChange",
        "onInput",
      ]
    }
  >
    <ModalHeader
      data-spec="upsert-character-modal-header"
      tag="h4"
      toggle={[MockFunction]}
      wrapTag="div"
    >
      Edit Character
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <TooltipForm
        Renderable="UpsertCharacterForm"
        characterToEdit={
          Object {
            "characterId": 1,
          }
        }
        handleInputChange={[Function]}
      />
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        color="primary"
        tag="button"
      >
        Submit Character
      </Button>
       
      <Button
        color="secondary"
        data-spec="upsert-character-modal-close-button"
        onClick={[MockFunction]}
        tag="button"
      >
        Cancel
      </Button>
    </ModalFooter>
  </AvForm>
</Modal>
`;

exports[`rendering should render valid snapshot if character is new 1`] = `
<Modal
  autoFocus={true}
  backdrop={true}
  backdropTransitionTimeout={150}
  data-spec="upsert-character-modal"
  fade={true}
  isOpen={true}
  keyboard={true}
  modalTransitionTimeout={300}
  toggle={[MockFunction]}
  zIndex={1050}
>
  <AvForm
    data-spec="upsert-character-modal-form"
    method="get"
    model={Object {}}
    onInvalidSubmit={[Function]}
    onKeyDown={[Function]}
    onSubmit={[Function]}
    onValidSubmit={[Function]}
    tag={[Function]}
    validationEvent={
      Array [
        "onChange",
        "onInput",
      ]
    }
  >
    <ModalHeader
      data-spec="upsert-character-modal-header"
      tag="h4"
      toggle={[MockFunction]}
      wrapTag="div"
    >
      Add Character
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <TooltipForm
        Renderable="UpsertCharacterForm"
        characterToEdit={Object {}}
        handleInputChange={[Function]}
      />
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        color="primary"
        tag="button"
      >
        Submit Character
      </Button>
       
      <Button
        color="secondary"
        data-spec="upsert-character-modal-close-button"
        onClick={[MockFunction]}
        tag="button"
      >
        Cancel
      </Button>
    </ModalFooter>
  </AvForm>
</Modal>
`;
